### ISSUE TITLE
IsRegistered throws IndexOutOfRangeException for open generic type.

### ISSUE THREAD

**Issue by hypercodeplace** [autofac/Autofac#1040]
Assembly version: 4.9.4.0

```C#
[Test]
public void IsRegisteredTest()
{
    void Test(Type typeFrom, Type typeTo)
    {
        var containerBuilder = new ContainerBuilder();
        containerBuilder.RegisterGeneric(typeTo).As(typeFrom);

        var container = containerBuilder.Build();

        Assert.IsFalse(container.IsRegistered(typeFrom));
    }

    Test(typeof(IFoo<>), typeof(Foo<>)); // Passed
    Test(typeof(Foo<>), typeof(Foo<>));  // Failed with "IndexOutOfRangeException".
}

public interface IFoo<T> { }
public class Foo<T> : IFoo<T> { }
```

The stack trace:

```
System.IndexOutOfRangeException : Index was outside the bounds of the array.
   at Autofac.Util.TypeExtensions.IsCompatibleWithGenericParameterConstraints(Type genericTypeDefinition, Type[] parameters)
   at Autofac.Features.OpenGenerics.OpenGenericServiceBinder.TryBindServiceType(Service service, IEnumerable`1 configuredOpenGenericServices, Type openGenericImplementationType, Type& constructedImplementationType, Service[]& constructedServices)
   at Autofac.Features.OpenGenerics.OpenGenericRegistrationSource.RegistrationsFor(Service service, Func`2 registrationAccessor)+MoveNext()
   at Autofac.Core.Registration.ComponentRegistry.GetInitializedServiceInfo(Service service)
   at Autofac.Core.Registration.ComponentRegistry.IsRegistered(Service service)
   at Autofac.ResolutionExtensions.IsRegisteredService(IComponentContext context, Service service)
   at Autofac.ResolutionExtensions.IsRegistered(IComponentContext context, Type serviceType)
```

**Comment by delepster** [comment]
Any hint on when the fix will be released?

**Comment by tillig** [comment]
It will be released in 5.0. There is no timeline set on 5.0. If you'd like to see a faster 5.0 release, helping out in the form of pull requests is the best way to accelerate the timeline.
